Support `RayTransform` custom backends

adriaangraas opened the pull request:
At this point it is difficult from the user level to specify a different backend for the `RayTransform` operator. (Current backends are Scikit-image, ASTRA CPU and ASTRA CUDA). Loading a custom backend is necessary if you like to modify the way ASTRA is called, for instance.

In this PR I propose to move some backend logic out of the `RayTransformBase` and `RayTransform` classes and into new `RayTransformImplBase` class and implementation classes. Something like this was already in place for `impl='astra_cuda'`. With this PR a user can for instance subclass the ASTRA CPU implementation and feed its type into the `impl` argument (strategy design pattern).

```python
    class MyRayTransformImpl(tomo.backends.AstraCpuRayTransformImpl):
        def call_forward(...):
            my_custom_stuff()

    ray_trafo = odl.tomo.RayTransform(reco_space, geometry, impl=MyRayTransformImpl)
```

Any implementation must subclass `RayTransformImplBase`, essentially forcing the `call_forward()` and `call_backward()`. There are some optional methods in the base class, such as `supports_geometry(geom)` and `supports_reco_space()`. 

Behavior and API for `RayTransform` and `RayBackProjection` remains backward compatible, in particular you can still specify `'astra_cpu`' etc. to `impl`. However, `AstraCudaProjectorImpl` and `AstraCudaBackProjectorImpl` are merged into `AstraCudaRayTransformImpl`. A lot of unnecessarily duplicated code has been removed there, and I didn't expect a lot of users to directly rely on these classes.

I'll make proper docs and fix the tests :) but I first wanted to send this in for some feedback on the concept.

pep8speaks commented:
Checking new PR...



* In the file [`odl/tomo/backends/astra_cpu.py`](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/backends/astra_cpu.py):

> [Line 276:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/backends/astra_cpu.py#L276): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 272:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/backends/astra_cpu.py#L272): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 206:13](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/backends/astra_cpu.py#L206): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 66:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/backends/astra_cpu.py#L66): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)

* In the file [`odl/tomo/backends/astra_cuda.py`](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/backends/astra_cuda.py):

> [Line 117:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/backends/astra_cuda.py#L117): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 77:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/backends/astra_cuda.py#L77): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 22:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/backends/astra_cuda.py#L22): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)

* In the file [`odl/tomo/backends/impl.py`](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/backends/impl.py):

> [Line 44:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/backends/impl.py#L44): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 40:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/backends/impl.py#L40): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)



* In the file [`odl/tomo/operators/ray_trafo.py`](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/operators/ray_trafo.py):

> [Line 502:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/operators/ray_trafo.py#L502): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 416:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/operators/ray_trafo.py#L416): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 263:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/operators/ray_trafo.py#L263): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 256:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/operators/ray_trafo.py#L256): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 255:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/operators/ray_trafo.py#L255): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 238:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/operators/ray_trafo.py#L238): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 227:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/operators/ray_trafo.py#L227): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 24:80](https://github.com/odlgroup/odl/blob/3567dc2ec6c51ad87816f418dbb5ffa1127dc002/odl/tomo/operators/ray_trafo.py#L24): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -285,9 +186,96 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+                        "slow for volumes of this size. Consider using ASTRA. "
+                        "This warning can be disabled by explicitly setting "
+                        "`impl='skimage'`.",
+                        RuntimeWarning)
```
Why this abstraction leak? In principle, the warning text could be part of the implementation class, so `RayTransformBase` wouldn't have to know about specific impls.

adriaangraas replied:
The motivation is to keep the same error messages, and a backend class is unaware of the others. The abstraction doesn't leak I believe, `RayTransform` is just aware of special cases. I changed the error messages a bit and moved this into the `__init__` of the backends as well.

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -396,41 +397,10 @@ def __init__(self, domain, geometry, **kwargs):
             variant='forward', **kwargs)
 
     def _call_real(self, x_real, out_real, **kwargs):
```
I think `_call_real` and `_call_complex` should also go into the implementation. That gives opportunity for optimization.

adriaangraas replied:
This I think is a very good point. I did this, but I'm not entirely sure that I picked the best solution yet. Since the `_call` is quite a big function I didn't want to duplicate it throughout the implementations. I added a `complexify()` function for now, let me know what you think.

kohr-h commented on odl/tomo/backends/astra_cpu.py:
```diff
@@ -29,6 +30,7 @@
     'astra_cpu_forward_projector',
     'astra_cpu_back_projector',
     'default_astra_proj_type',
+    'AstraCpuRayTransformImpl',
```
We shouldn't publicly expose these.

adriaangraas replied:
Indeed! Thanks.

adriaangraas commented on odl/tomo/backends/impl.py:
```diff
@@ -0,0 +1,73 @@
+__all__ = ('RayTransformImplBase',)
+
+
+class RayTransformImplBase(ABC):
```
Good to know!

adriaangraas commented on odl/tomo/backends/impl.py:
```diff
@@ -0,0 +1,73 @@
+    @staticmethod
+    def can_handle_size(size):
+        """A very general way to check if the implementation is capable
+        handling reconstruction volumes of a given size."""
```
I moved this now into the `__init__`.

adriaangraas commented on odl/tomo/backends/impl.py:
```diff
@@ -0,0 +1,73 @@
+    def can_handle_size(size):
+        """A very general way to check if the implementation is capable
+        handling reconstruction volumes of a given size."""
+        return True  # by default no assumptions are made
```
Now irrelevant since the base class is removed.

adriaangraas commented on odl/tomo/backends/skimage_radon.py:
```diff
@@ -171,3 +174,47 @@ def skimage_radon_back_projector(sinogram, geometry, vol_space, out=None):
+    def supports_geometry(cls, geometry):
+        if not isinstance(geometry, Parallel2dGeometry):
+            return TypeError("{!r} backend only supports 2d parallel "
+                             'geometries'.format(cls.__name__))
```
This indeed smells and I didn't really like it as well. I moved the checking logic into the `__init__` as well. There it makes sense to raise errors. This makes it impossible for the composing class to check ahead of time if the backend supports the geometry, but I think cleaner code is way more preferable anyway.

adriaangraas commented:
Hi @kohr-h, thanks for reviewing!

I think you have fair points here. Although I like programming to interfaces (more to open the class for extension, not necessarily to standardize), I agree that OOP machinery easily ends up with a lot of boilerplate code. I tried removing the `RayTransformImplBase`, and I think that goes fairly well with just a few concessions: there is small bit of code duplication in the `__init__` of each implementation, and I think it'd have been a good place to implement complex `_call`s. I added a few runtime checks in `RayTransform` verifying the existence of `call_forward` and/or `call_backward`. All in all I think it simplified the code quite a bit. 

I also removed the `RayTransformBase` class, as you suggested. The `RayTransform`, and `RayBackProjection` are now directly derived from `Operator`. I completely agree with the former solution being suboptimal: subclassing, to have two `variants`, while simultaneously having the same variance in the forward and `adjoint()` is very confusing. Also, I don't think subclassing was the right pattern there to begin with. I'm not sure how you'd design the inner `adjoint` class, but I thought that a very natural way was to compose `RayBackProjection` _using_ instead of _extending_ a `RayTransform`. I gave this a try as well, let me know what you think.

I've two direct questions:
1. Now that I moved warnings into the implementations I cannot suppress the `size` warning, [like here](https://github.com/odlgroup/odl/blob/9382c4e290facde061cfda9ee8aa639a63745412/odl/tomo/backends/astra_cpu.py#L279). I could either add a `verbose` flag or try to catch the warnings, but neither seem very appealing. Do you have a good solution here?
2. I used a `complexify` function to turn real-valued functions into complex-valued functions. Is there a nicer way, maybe using ODL utilities, to achieve this?

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,164 +161,119 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+        impl = kwargs.pop('impl', None)
+        impl_type, self.__cached_impl = self._check_impl(impl)
+        self._impl_type = impl_type
+        self.__impl = impl.lower() if isinstance(impl, str) else impl_type
```
This I find a bit worrying. We try very hard to make every stored attribute a fixed type, or `None` if it makes sense, in particular for those that back public properties. Never one of several possible types. That's a recipe for downstream bugs.
I'd suggest you keep `impl` to always be a `str` and look up the corresponding type in a dictionary when needed. I don't think there's a need to store the type in the class. The fewer attributes, the better.

adriaangraas replied:
I changed `RayTransform.impl` so that it always returns a `str`. The reason why we currently have this hassle of storing the type is that `RayTransform` has "knowledge" of how to instantiate the implementation. Normally I wouldn't like that, but it makes sense because `RayTransform` needs to reinitiate the backend when `use_cache=False`. I guess I can also change the class so that it will directly use `type(self.__cached_impl)`, and then we'd have to make sure that there is always a `self.__cached_impl` available.

kohr-h replied:
Hm, yes, that makes sense. I'll have a closer look.

I recognize this problem from classes like `DiscreteLp` that are composed of other classes. The moment you want to "convert" the containing class, e.g., change the `dtype`, it suddenly needs to know how to instantiate its components. An annoying issue, and I haven't found a good solution yet.

adriaangraas replied:
Thought about this. To get "creation knowledge" outside of the class we'd need either DI (not very ODL-style) or push it creation off to some external factory class. A bit of an overkill. Since `RayTransform` really composes and not "aggregates" `Impl`s, I don't think it is too weird for `RayTransform` to instantiate the `Impl`s. However, I don't think I'm a big fan of the root of this problem, the `use_cache` methodology. Caching/recreating the `Impl` is a way to prevent excessive memory usage after retrieving the volume data from GPU, while guaranteeing a `DiscreteLp` element. But do we not need to give users more explicit control over what is on the GPU in some future anyway? Like neural networks, that often implement `.numpy()` methods. And, if CuPy makes it into ODL, some methods may return CuPy arrays which probably shouldn't be cast into Numpy arrays by default.

At the moment I find only one example using `use_cache`. I'd think that we could either leave cache out, and people that need this can still `del ray_trafo` or reset `ray_trafo.impl`. Or maybe we'll leave it in until we find a good solution to the problems above?

kohr-h replied:
These are all good thoughts and important to keep in mind. For the moment, though, I think the current solution is (more than) acceptable. One of the points of this PR is to make the `RayTransform` extensible by using a common interface for the `Impl` classes, which makes it easy to instantiate them correctly in the caching logic.
Regarding future-proofness of the caching: ASTRA (as of now) works with host memory via copying or device memory via linking, and with the current mechanism we have the first case covered. At some point we'll modify the caching logic to account for existing GPU memory. That will require a bit of restructuring, but I don't see more cases than that.

adriaangraas replied:
I agree that this is not for this PR :)

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,164 +161,119 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+        self._impl_type = impl_type
+        self.__impl = impl.lower() if isinstance(impl, str) else impl_type
+
+        self.geometry = geometry
```
This should ultimately be a `@property`, too.

adriaangraas replied:
I did this, but why is this exactly? For autocompletion?

kohr-h replied:
The main reason is to make our objects "publicly immutable". You can't assign to properties unless a setter is defined, whereas the plain attributes can be changed arbitrarily. It's not perfect but at least conveys an intention.

adriaangraas replied:
That makes a lot of sense, thanks!

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,164 +161,119 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
-    def impl(self):
-        """Implementation back-end for the evaluation of this operator."""
-        return self.__impl
+    def _check_impl(self, impl):
```
This could be a `@staticmethod`.

adriaangraas replied:
Changed this.

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -499,73 +345,51 @@ def __init__(self, range, geometry, **kwargs):
-                                     use_cache=self.use_cache,
-                                     **kwargs)
-        return self._adjoint
+        return copy.copy(self._ray_trafo)
```
See my suggestion for an inner class. It's largely the same, just without copying (which I don't think is necessary anyway).

adriaangraas replied:
I like this. Makes `RayTransform` really clean and expressive. I was thinking that perhaps `.adjoint` had to return a new class (as it was before), but I don't think that is sensible with an `@property`.

adriaangraas commented on odl/tomo/backends/astra_cpu.py:
```diff
@@ -11,11 +11,13 @@
 
 from odl.discr import DiscreteLp, DiscreteLpElement
-from odl.tomo.backends.astra_setup import (
+from odl.tomo.backends import (
```
Thanks! Good catch, not sure why I changed that.

adriaangraas replied:
Thanks, I missed this.

adriaangraas commented on odl/tomo/backends/astra_cpu.py:
```diff
@@ -200,8 +200,8 @@ def astra_cpu_back_projector(proj_data, geometry, vol_space, out=None,
-                         'geometry do not match'.format(
-                             vol_space.ndim, geometry.ndim))
+                         'geometry do not match'.format(vol_space.ndim,
+                                                        geometry.ndim))
```
Hm, I generally like to not have a end parenthesis on a newline, saving a line. Other than that I agree.

kohr-h replied:
That's a matter of taste, sure. For other projects I nowadays use the `black` code formatter,  and it prefers the format I show above. Sometimes it's useful when you have chained calls, like 

    np.some_function(
        very_long_expression_that_fills_a_whole_line_or_more
    ).astype("float32")

But never mind.

adriaangraas commented on odl/tomo/backends/astra_cpu.py:
```diff
@@ -249,6 +249,65 @@ def astra_cpu_back_projector(proj_data, geometry, vol_space, out=None,
     return out
 
 
+class AstraCpu:
```
Fixed that! I need to properly check docs later.

adriaangraas commented on odl/tomo/backends/astra_cuda.py:
```diff
@@ -59,9 +60,39 @@ def __init__(self, geometry, reco_space, proj_space):
+            mid_pt = geometry.det_params.mid_pt
+            for i, angle in enumerate(geometry.angles):
+                if abs(np.dot(axis,
+                              geometry.det_to_src(angle, mid_pt))) < 1e-4:
```
Agreed!

adriaangraas commented on odl/tomo/backends/astra_cpu.py:
```diff
@@ -249,6 +249,65 @@ def astra_cpu_back_projector(proj_data, geometry, vol_space, out=None,
+        ----------
+        geometry : `Geometry`
+            Geometry defining the tomographic setup.
+        reco_space : `DiscreteLp`
```
Changed this everywhere. Hope nobody is calling `ray_trafo = RayTransform(reco_space=...)`?

kohr-h replied:
I guess `reco_space` was so long and hard to type that nobody bothered and just used a positional argument. Some of our own examples might do that, though.

adriaangraas commented on odl/tomo/backends/astra_cuda.py:
```diff
@@ -165,70 +145,78 @@ def create_ids(self):
-            aproj.delete(self.proj_id)
-            self.proj_id = None
+    def call_forward(self, x, out, **kwargs):
+        return complexify(self._call_forward_real, self.reco_space) \
```
Irrelevant now with the decorator.

adriaangraas commented on odl/tomo/backends/skimage_radon.py:
```diff
@@ -11,13 +11,19 @@
-from odl.discr import uniform_discr_frompartition, uniform_partition
+from odl.tomo import Parallel2dGeometry, Geometry
+
+from odl.discr import uniform_discr_frompartition, uniform_partition, \
```
Done.

adriaangraas commented on odl/tomo/backends/astra_cpu.py:
```diff
@@ -249,6 +249,65 @@ def astra_cpu_back_projector(proj_data, geometry, vol_space, out=None,
     return out
 
 
+class AstraCpu:
```
I was thinking this might clash with the ASTRA naming for a _Projector_, which is either a forward projector or a backprojector, while our class does both. I now named this `<backend>Impl`. What do you think?

kohr-h replied:
Ah yes, that's fine. `Impl` sounds so obscure that nobody will dare to use these classes directly :wink:

adriaangraas commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -19,33 +20,34 @@
+from odl.tomo.geometry import Geometry
+
+# Backends that are implemented in ODL and can be specified via `impl`
+_SUPPORTED_IMPL = {
```
Sounds good.

adriaangraas commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -19,33 +20,34 @@
-class RayTransformBase(Operator):
-
+class RayTransform(Operator):
     """Base class for ray transforms containing common attributes."""
```
Fixed!

adriaangraas commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -19,33 +20,34 @@
     """Base class for ray transforms containing common attributes."""
 
-    def __init__(self, reco_space, geometry, variant, **kwargs):
+    def __init__(self, reco_space, geometry, **kwargs):
```
And everywhere else :)

adriaangraas commented:
Thanks again @kohr-h, I don't mind the nitty comments :+1: 

Open remain mostly decisions from the unresolved conversations above. Is there anything else that would be good to add to this PR?

About the decorator: Although the syntax is really clean in the class I don't like that it implicitly passes `self` through the call. I would have liked it better when it was more restricted to its input, or more widely applicable. I'll give it a thought.

In a next update I'll try to have a look at the docs and tests!

kohr-h commented on odl/tomo/backends/util.py:
```diff
@@ -0,0 +1,26 @@
+
+    def wrapper(self, x, out, **kwargs):
+        if self.vol_space.is_real and self.proj_space.is_real:
+            fn(self, x, out, **kwargs)
```
You're correct, `self` has to be supplied. Same for the complex case below.

adriaangraas replied:
Sharp! Real-time feedback :)

kohr-h replied:
Yes, but now it's late enough and my sharpness declines rapidly :-)

kohr-h commented:
> Open remain mostly decisions from the unresolved conversations above. Is there anything else that would be good to add to this PR?

I'll go through it in detail again to get a better picture. But one thing that comes to my mind is the `impl` string. Maybe instead of using `__class__.__name__`, the `...Impl` classes should have a class attribute `impl_name` or so, and register themselves with that name in a global dictionary in `ray_trafo.py`. That smells a bit like danger of circular imports, but I think if the code in `ray_trafo.py` knows nothing about the specific `Impl`'s it shouldn't even have to import any of those modules.

Do you think that's reasonable?

(Or maybe that's already the current status.)

adriaangraas commented:
Hi @kohr-h. Hm, I don't know about the global dictionary, sounds like code that is hard to read and debug. I can see that this allows users to register their only class to the global register, which will then automatically be used (if the priorities go right) by `RayTransform`. I think the expressive `RayTransform(impl=MyImpl())` is way less magical and achieves the same thing. Furthermore `RayTransform` is already unaware of the default implementations, because it reads from `_AVAILABLE_IMPLS`. Maybe I am missing something here?

adriaangraas commented on odl/tomo/backends/util.py:
```diff
@@ -4,18 +4,17 @@
 
             if out is None:
-                out = range.element()
+                out = self.proj_space.element()
```
I need to fix this, maybe write a test to cover this?

kohr-h replied:
Yes, that's a good idea.

adriaangraas replied:
Fixed this and updated the complex test to cover this.

kohr-h commented:
> _Update_: I guess we could make `_AVAILABLE_IMPLS` and `_IMPL_STR2TYPE` public though, so that amending at runtime is possible.

That's what I mean. Two things, then: maybe as a public dictionary, the name should be a bit more verbose. And we don't need the two above, `_AVAILABLE_IMPLS` could just be `_IMPL_STR2TYPE.keys()`.

kohr-h commented on odl/tomo/backends/skimage_radon.py:
```diff
@@ -171,3 +177,79 @@ def skimage_radon_back_projector(sinogram, geometry, vol_space, out=None):
     return out
+
+
+class SkimageImpl:
```
Suggestion: `SkImageImpl` (first "I" capitalized)

adriaangraas replied:
Good!

adriaangraas commented:
I've changed `_IMPL_STR2TYPE` to an `OrderedDict` named `RAY_TRAFO_IMPLS`. Maybe verboseness is good here when the variable goes public? I'm also fine with the other name. The last inserted item now automatically has the highest priority.
I didn't manage to remove the second global without introducing much extra code. `RayTransform` needs to have different lists in order to tell the user that an `impl` exists but is not installed. That list can remain private though.

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+                                           range=proj_space, linear=True)
 
+    @staticmethod
+    def _check_impl(impl):
```
This guy does a lot more than mere checking. Maybe `_initialize_impl`?

adriaangraas replied:
Sounds good.

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
 
-    def __init__(self, range, geometry, **kwargs):
-        """Initialize a new instance.
+    def create_impl(self, use_cache=True):
+        """Fetches or instantiates implementation backend for evaluation.
```
Those two cases are perhaps best summarized by the name `get`.

adriaangraas replied:
For me that reminds me of the dumb getters/setters of many other languages, but that might not be a valid reason. Changed it.

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
-
-            For the default ``None``, the fastest available back-end is
-            used, tried in the above order.
+        bool : use_cache
```
Please mirror at the colon

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
-            For the default ``None``, the fastest available back-end is
-            used, tried in the above order.
+        bool : use_cache
+            If ``True`` returns the cached implementation backend, if it
```
```suggestion
            If ``True``, return the cached implementation backend, if it
```

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+        bool : use_cache
+            If ``True`` returns the cached implementation backend, if it
+            was generated in a previous call (or given with ``__init__``).
+            If ``False`` a new instance of the backend will be generated,
```
```suggestion
            If ``False``, a new instance of the backend will be generated,
```

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
-            reco_space=range, proj_space=domain, geometry=geometry,
-            variant='backward', **kwargs)
+    def _call(self, x, out=None, **kwargs):
+        """Forward projection
```
```suggestion
        """Forward projection.
```

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+        Parameters
+        ----------
+        x : DiscreteLpElement
+            A volume. Must be an element from `RayTransform.domain`.
```
```suggestion
            A volume. Must be an element of `RayTransform.domain`.
```

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+        ----------
+        x : DiscreteLpElement
+            A volume. Must be an element from `RayTransform.domain`.
+        out : `RayTransform.range` element, optional
+            A DiscreteLpElement from `RayTransform.range`, to which the result
+            of the operator evaluation is written.
```
```suggestion
        out : `RayTransform.range` element, optional
            Element to which the result of the operator evaluation is written.
```
I think this doc can be quite minimal, it's only for devs and does not deviate a lot from `Operator._call` anyway. I'd also be okay with a one-liner.

adriaangraas replied:
I'll keep that in mind also for other docs.

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+            A DiscreteLpElement from `RayTransform.range`, to which the result
+            of the operator evaluation is written.
+        **kwargs
+            Arbitrary keyword arguments, passed on to the implementation
+            backend.
```
```suggestion
            Extra keyword arguments, passed on to the implementation
            backend.
```
"Arbitrary" goes a bit too far IMO

adriaangraas replied:
Agreed. Changed at two places.

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
-                # Should never happen
-                raise RuntimeError('bad `impl` {!r}'.format(self.impl))
+        return self.create_impl(self.use_cache) \
+            .call_forward(x, out, **kwargs)
```
```suggestion
        return self.create_impl(self.use_cache).call_forward(
            x, out, **kwargs
        )
```

adriaangraas replied:
Recently tried to stop with the backslash habit, but with `get_impl` the line fits!

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+        Returns
+        -------
+        DiscreteLpElement
+            A sinogram, from the projection space `RayTransform.range`.
```
Not every ray transform results in a (single) sinogram. Maybe keep it generic, like "Result of the projection" or so.

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+                """Adjoint of the discrete Ray transform between L^p spaces."""
+
+                def _call(self, x, out=None, **kwargs):
+                    """Backprojection
```
```suggestion
                    """Backprojection.
```

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+                    -------
+                    DiscreteLpElement
+                        The reconstructed volume, an element in
+                        `RayBackProjection.range`.
```
See above

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+                    """
+
+                    return ray_trafo.create_impl(ray_trafo.use_cache) \
+                        .call_backward(x, out, **kwargs)
```
```suggestion
                    return ray_trafo.create_impl(ray_trafo.use_cache).call_backward(
                        x, out, **kwargs
                    )
```

adriaangraas commented:
Thank you @kohr-h! Going to pick this up very soon (days).

ozanoktem commented:
I want to thank both @adriaangraas and @kohr-h for the effort in providing new functional to ODL. Your work is much appreciated.

kohr-h commented:
@adriaangraas I'm currently rebasing on master. There has been a rather big internal API change (#1459) that impacts this PR, so I think it's easiest if I deploy the necessary changes here so you can focus on the actual content of the PR. Is that okay? So don't worry about an upcoming force-push :wink:

kohr-h commented:
Okay, I pushed a couple of commits: the rebase, fixes of rebase accidents and some formatting updates. Please take it from here @adriaangraas

adriaangraas commented on odl/tomo/backends/astra_cuda.py:
```diff
@@ -10,117 +10,107 @@
 from packaging.version import parse as parse_version
 
 from odl.discr import DiscreteLp
+from odl.tomo.backends.util import _add_default_complex_impl
```
Thanks! I've now checked a box for sorting imports, I hope that is consistent with `isort`, otherwise I'll configure that.

adriaangraas commented on odl/tomo/backends/astra_cuda.py:
```diff
@@ -10,117 +10,107 @@
 
         self.create_ids()
 
         # ASTRA projectors are not thread-safe, thus we need to lock ourselves
```
Haha true, the scope of this PR was creeping a bit anyway :smile:

adriaangraas commented on odl/tomo/backends/skimage_radon.py:
```diff
@@ -11,13 +11,19 @@
+import warnings
 
-from odl.discr import uniform_discr_frompartition, uniform_partition
+from odl.tomo import Parallel2dGeometry, Geometry
```
Missed this, you mean to import directly via `odl.discr.discr_space`? Here we are in `odl.tomo.backends`?

kohr-h replied:
No, I mean the `tomo` imports. Maybe this is based on me misunderstanding the Python import system, but it seems to me that it would be easier to break import cycles when import paths are deeper.
The above version is actually a cyclic import, because it says "import the `odl.tomo`, subpackage, *which includes myself*, and give me its attributes `Parallel2dGeometry` and `Geometry`.

adriaangraas replied:
I guess it makes sense not to import from your parent indeed. Wonder what happens though at import-time: if the parent is already imported, then Python may not re-import the parent again? :thinking:

adriaangraas commented on odl/tomo/backends/util.py:
```diff
@@ -0,0 +1,44 @@
+    This function helps `RayTransform` implementations by splitting
+    complex-valued forward/backward function calls into two real-valued calls.
+
+    Parameters
+    ----------
+    fn : Callable
+        Function with signature ``fn(self, x, out=None, **kwargs)``.
+        ``self`` must be an object instance having ``self.vol_space`` and
+        ``self.proj_space``. These spaces must be both real or both complex.
+
+    Returns
+    -------
+    Callable
+        A wrapper function to be executed by a Python decorator.
```
Ok! I've changed this a bit, let me know if this is what you meant.

adriaangraas commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+        impl_type, self.__cached_impl = self._check_impl(impl)
+        self._impl_type = impl_type
+        self.__impl = impl.lower() \
+            if isinstance(impl, str) else impl_type.__name__
```
Agreed! Did your formatter fix this (is that `black`?). Maybe I should set that up as well...

kohr-h replied:
I use `black` privately, but not on ODL - I just use its suggested style. Using the tool would be a huge time-saver, but also a huge commit that screws up Git history badly. Maybe we can do it at some point anyway :-)

adriaangraas replied:
A quick search finds that this is possible without destroying Git history, but at the cost that every developer needs to re-clone their repository. Seems reasonable.

adriaangraas commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+
+                if impl.lower() not in RAY_TRAFO_IMPLS.keys():
+                    raise ValueError(
+                        '{!r} back-end not available'.format(impl))
```
Maybe this would have been better as a class constant, but I'm also in favor of simpler code. Removed it.

adriaangraas commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
-                        self._astra_wrapper = astra_wrapper
+        else:
+            # User did specify `impl`
+            if isinstance(impl, str):
```
Did that. Note that there is a `.lower()` required later in the class though. Was there ever a use case for this, curious about it.

kohr-h commented on odl/tomo/backends/util.py:
````diff
@@ -0,0 +1,71 @@
+
+    The wrapper will only work for methods of which the class provides a
+    `vol_space` and `proj_space`. Typically, this will then work as a decorator
+    on the method, e.g.:
+    ```
+        @_add_default_complex_impl
+        def call_forward(self, x, out=None, **kwargs):
+            # Code that should run for real input and output
+    ```
````
```suggestion
    on the method, e.g. ::

        @_add_default_complex_impl
        def call_forward(self, x, out=None, **kwargs):
            # Code that should run for real input and output
```

Note that the docstrings are parsed as reStructuredText, not Markdown. So the triple-quote thing won't work.

kohr-h commented on odl/tomo/operators/ray_trafo.py:
```diff
@@ -273,299 +163,225 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+            if is_string(impl):
+                if impl.lower() not in RAY_TRAFO_IMPLS.keys():
+                    raise ValueError(
+                        'The {!r} `impl` is not found. This `impl` is either '
+                        'not supported, it may be misspelled, or external '
+                        'packages required are not available. Consult '
+                        '`RAY_TRAFO_IMPLS` to find the run-time available '
+                        'implementations.'.format(impl)
```
Great error message!

adriaangraas replied:
Thanks :D

kohr-h commented:
> One little thing

And I just did it myself :-)

kohr-h commented:
Not waiting for CI. The last commit built fine, so I'll just merge it.

kohr-h commented:
Thanks a lot for all the work @adriaangraas!

adriaangraas commented:
And thanks a ton for all the guidance @kohr-h, I've also learned a lot!
